source/language/gates.rst

```diff
@@ -233,24 +236,83 @@ of the gate definition.
      U(0, 0, 0) a[0];
    }
 
-Only built-in gate statements, calls to previously defined gates, and
-timing directives can appear in ``body``. For example, it is not valid to
-declare a classical register in a gate body. Looping constructs over these quantum
-statements are valid.
+'General' gate declarations have a similar structure, with minor differences:
 
-The statements in the body
-can only refer to the symbols given in the parameter or argument list,
-and these symbols are scoped only to the subroutine body.
+.. code-block:: c
 
-An empty body corresponds to the identity gate.
+   // comment
+   gate name(typedParams) : typedQargs
+   {
+     body
+   }
 
-Gates must be declared before use and
-cannot call themselves. The statement ``name(params) qargs;`` applies the gate,
-and the variable parameters ``params`` can have any numeric type.
+The (again, optional) parameter list ``typedParams`` is a comma-separated list of variable
+parameters, which in this case must be provided with explicit type specifications.
+The list ``typedQargs`` is a comma-separated list of operands which are either individual
+qubits, or registers/arrays of qubits, where each is again provided with an explicit type
+specification. 
+If there are no variable parameters, the parentheses are optional. At least
+one quantum operand is required, and the quantum operands
+must be immediately preceded by a ``:`` delimiter.
+The arguments in ``typedQargs`` can be indexed within the body
+of the gate definition if, and only if, it is a register or array of qubits.
+
+.. code-block:: c
+
+   // this is ok:
+   gate g (angle alpha, int k): qubit[3] a
+   {
+     U(0, 0, alpha) a[0];
+     U(0, 0, alpha/k) a[1];
+     U(0, 0, alpha/(k**2)) a[2];
+   }
+
+   // this is also ok:
+   gate qutritX : qubit[2] a {
+     x a[1];
+     cx a[1], a[0];
+     cx a[0], a[1];
+   }
+
+   // this is invalid:
```

Comment:
would be useful to call out why it is invalid

Reply:
Noted, I've amended this example and the previous one for the ``gate`` with the OpenQASM2 declaration syntax.